src/fetching.rs

```diff
@@ -109,7 +138,25 @@ impl<T, C> Fetcher<T, C> {
             }
 
             // Now we are responsible for fetching the object, reach out to the provider.
-            let res = provider.fetch(req).await;
+            let res = loop {
+                if let Some(res) = provider.fetch(req).await {
+                    break res;
+                }
+
+                // We only fetch objects which are known to exist, so we should eventually succeed
+                // in fetching if we retry enough. For example, we may be fetching a block from a
+                // peer who hasn't received the block yet.
+                //
+                // To understand why it is ok to retry indefinitely, think about manual
+                // intervention: if we don't retry, or retry with a limit, we may require manual
+                // intervention whenever a query service fails to fetch a resource that should exist
+                // and stops retrying, since it now may never receive that resource. With indefinite
+                // fetching, we require manual intervention only when active fetches are
+                // accumulating because a peer which _should_ have the resource isn't providing it.
+                // In this case, we would require manual intervention on the peer anyways.
+                tracing::warn!("failed to fetch {req:?}, will retry in {retry_delay:?}");
+                sleep(retry_delay).await;
```

Comment:
In the future it might be nice to have a backoff setting

Reply:
Good idea, I implemented a simple exponential backoff bounded by a maximum (which is configurable), this lets us try a few times very quickly but then back off to a very slow retry rate to reduce spam

Reply:
nice!